blog: Developer Opportunities at the Monero Project (#2119)
monero-project/monero-site: _posts/2023-01-01-seraphis-jamtis-developer-opportunities.md

Over time many features and privacy enhancements were added, and today the code has grown to maybe something like tripple original size. Given how many different people worked on the code for many years already, and given the true open source style of development where nobody can just dictate things, the code is still in reasonable shape. There are nevertheless parts with a size and a complexity that makes them struggle to fit within the original framework that still largely stands.

We have [developer documentation](https://www.getmonero.org/resources/developer-guides/), but unfortunately there is nothing that documents the overall code structure, or the hierarchy of classes, or gives a high-level overview of the functionality. This together with the fact that C++ is not the easiest of programming languages when it comes to understand a large codebase makes it a bit difficult for new devs to find their way around; but on the other hand the "culture" among the devs is very welcoming and supporting to people who are serious to contribute.

> You find the project [here on GitHub](https://github.com/monero-project). The most important repository is the one with the core software, e.g. the Monero *daemon* that spans up the network and manages the blockchain, see [its repository](https://github.com/monero-project/monero). You find the "places to be" like the dev-oriented Libera.chat IRC channel *#monero-dev* on the ["Hangouts" page](https://www.getmonero.org/community/hangouts/).

Reviewer: Not sure if "here on github" is the best phrasing. People will be seeing the post on the website. Maybe just better to simply use "on Github"

Author: Agreed, changed.

### Seraphis and Jamtis

> A group of Monero devs started to work on two brand-new and sizeable new technologies called *Seraphis* and *Jamtis* which will improve the privacy of Monero transactions further and will solve many problems regarding current functionality. Those devs get support from the inventor of Seraphis, [UkoeHB](https://github.com/UkoeHB), and the inventor of Jamtis, [Tevador](https://github.com/tevador). The resulting cryptocurrency will be quite different from today's Monero, although existing users will be offered a seamless upgrade path, and funds from all the way back to 2014 remain fully spendable indefinitely.

Reviewer 2: > brand-new and sizeable new

could this be reworded without repeating new?

Author: Yeah, the second "new" is simply redundant :)

Over the course of about one year UkoeHB built a library that implements all the cryptography behind Seraphis and contains code to create transactions. You find it [in UkoeHB's repository](https://github.com/UkoeHB/monero/tree/seraphis_lib/src/seraphis).

Work has now started on the so-called *Seraphis wallet* that will make heavy use of the Seraphis library. *wallet* in this context does not mean an app that end users run to manage their XMR, but that component of the Monero core code that such apps will build on. The main functionality of that component: Reading and writing wallet files that contain keys and all owned outputs, scanning the blockchain for incoming transactions, building new transactions and submitting them to the network.

> The current plan is to fully replace *wallet2*, the existing component doing all this. This will require a lot of adjustments in the rest of the Monero core codebase, and of course reworking all wallet apps that depend on it. For more details [check here](https://github.com/seraphis-migration/strategy/wiki).

Author: Agreed again, changed.

> The current plan is to fully replace the current component with this role called *wallet2*. This in turn will make many adjustments in the rest of the Monero core codebase necessary, and of course reworking all wallet apps to use the new component. You find more details [here](https://github.com/seraphis-migration/strategy/wiki).

Reviewer 2: i struggled a bit reading this, consider my half-baked suggestion:
```suggestion
Currently, "wallet2" is handling all these roles and the plan is to replace it. This will require a lot of adjustments in the rest of the Monero core codebase (and of course, reworking all wallet apps that depend on it). You find more details [here](https://github.com/seraphis-migration/strategy/wiki).
```

Author: My attempt, a mix of your proposal with my existing paragraph:

> The current plan is to fully replace *wallet2*, the existing component doing all this. This will require a lot of adjustments in the rest of the Monero core codebase, and of course reworking all wallet apps that depend on it. For more details [check here](https://github.com/seraphis-migration/strategy/wiki).

Basically anybody can write a proposal for it, and if it gets positive feedback it moves to the funding stage where of course success depends on whether donators will find the proposed work and goal worthwhile.

Realistically however it's pretty hard to get a proposal accepted if you are all new to Monero, nobody in the community knows you yet and you have never contributed anything so far. Much is about trust and demonstrated abilities, as it typically is the case in open source projects.

The CCS has a long history of successfully funded and executed tasks.

Another option for funding Monero related work is the [MAGIC Monero Fund](https://magicgrants.org/funds/monero). Its nature is quite different from the CCS as it's a USA-registered 501(c)(3) charity. This enables tax-deductible (for citizens of the USA) donations in a variety of methods, and the option of USD-denominated payouts. There is also an additional requirement that fund recipients must verify their identity. The CCS allows its fund recipients to remain anonymous, yet only handles XMR.

### Conclusion

> There will be a lot of interesting work for devs that want to join the Monero project, for a long time. It may all look a bit unusual as a "place" to work at, and "get rich quick" isn't, but on the other hand think about it: Where else can you make a difference for a multi-billion-dollar venture that may become important in the future and make a difference, whoever you are, from wherever you are, regardless of education, only based on your knowledge and your willingness to make useful contributions?

Reviewer 2: > and "get rich quick" isn't 

hm.. if - "but it isn't" is used: "but on the other hand" would have to be changed.

Author: Because of your comment I noticed that "get rich quick" is the wrong metaphor to use because that hints at scams or at least shady things. That's not what I want to say here however, thus I arrived at the following proposal:

> It may be a bit unusual as a "place" to work at, and for getting rich probably look elsewhere, but on the other hand think about it:

---
layout: post
title: Developer Opportunities at the Monero Project
summary: Join the Monero dev community to implement the two brand-new technologies Seraphis and Jamtis 
tags: [community, development, seraphis, jamtis, ccs, magic]
author: René Brunner (rbrunner7)
---

> ### TL;DR

Author: I chose *Overview*.

> Monero is a cryptocurrency started in 2014. It's often called a *privacy coin*: Transaction senders, receivers and amounts are all not visible to outsiders on the blockchain, in stark contrast to most other cryptocurrencies, including Bitcoin. However, many proponents do not primarily see privacy as an end in itself but something that is needed for Monero for being able to function as a true electronic version of cash. For an ELI5 check e.g. [What is Monero](https://www.getmonero.org/get-started/what-is-monero/).

Author: I settled on "For an introduction see ..."

The Monero dev community is currently implementing two brand-new technologies called *Seraphis* and *Jamtis* to improve privacy further and add functionality. According to some still very rough estimates those may be ready to bring into service with a hardfork in 2 years' time. Quite a lot of new code needs to get implemented, and a lot of existing code modified and adjusted; it's probably the biggest single coding related project for the currency ever.

Does that sound like something you may want to work on? Can you write sophisticated C++ code, or are you able to learn to do so "on the job", and ready to work as a member of an open source dev team? Then please have a closer look.

> There is no company behind Monero to manage dev work; it's an open source project as true as they get nowadays. There are no VC companies involved in financing either that could force the project into directions that benefit themselves more than the Monero users. "Meritocracy" may be pretty grand and a bit overused as a concept, but if your work has merit you will get your code merged into software that is used to manage a cryptocurrency with a market valuation in the billions of USD.

Author: Quite in general, I see this text as quite a bit different than many other blog posts because its aim is not merely to inform, but it's sort of an advert, and it tries to make Monero look attractive not only as a project per se, but attractive to devs on the lookout for a new project to work on.

That's my motivation to mention those billions of USD: I want to make patently clear that Monero is not just one of the hundreds of small coins that are desperate for devs, but one of the most important and biggest ones. And there, even if we might not like it too much, current market valuation shows that.

> There is no company behind Monero to manage dev work; it's an open source project as true as they get nowadays. There are no VC companies involved in financing either that could force the project into directions that benefit themselves more than the Monero users. "Meritocracy" may be pretty grand and a bit overused as a concept, but if your work has merit you will get your code merged into software that is used to manage a cryptocurrency with a market valuation in the billions of USD.

Reviewer: > even if we might not like it too much, current market valuation shows that.

I disagree with this, as market value is easily gamed by centralized exchange and i don't consider it a useful metric, but that is only my opinion and i understand you prefer to keep it as it is.

> ### Project Background

Reviewer: Maybe this section could be trimmed down and the title changed.
At the end we are on the getmonero website where people can find a lot of info about the project background. Better to (at least partially and where possible) link to those and make the post shorter and more focused.

Author: In the "advert style" I mentioned above I would rather like those "talking points" to stand, and let them be directly visible in the text itself. You are right, these are certainly to be found and all nicely explained elsewhere on the website, but I try to bring my points home as directly as possible.

I also hope that if people do continue to read after they consumed the *Summary* they will give me some slack when it comes to text length.

> Monero came into existence 2014 when a group of 7 devs forked the code of a cryptocurrency called *Bytecoin* to start a new, independent blockchain with a fresh genesis block. Its starting point was therefore a moderately large C++ codebase built from scratch since maybe 2011 to implement a new privacy coin technology called *CryptoNote* for the first time.

Reviewer 2: ```suggestion
Monero came into existence in 2014 when a group of 7 devs forked the code of a cryptocurrency called *Bytecoin* to start a new, independent blockchain with a fresh genesis block. Its starting point was therefore a moderately large C++ codebase built from scratch since maybe 2011 to implement a new privacy coin technology called *CryptoNote* for the first time.
```

Author: Corrected.

Over time many features and privacy enhancements were added, and today the code has grown to maybe something like tripple original size. Given how many different people worked on the code for many years already, and given the true open source style of development where nobody can just dictate things, the code is still in reasonable shape. There are nevertheless parts with a size and a complexity that makes them struggle to fit within the original framework that still largely stands.

> Unfortunately there is nothing that properly documents the overall code structure, or the hierarchy of classes, or gives an overview of the functionality. This together with the fact that C++ is not the easiest of programming languages when it comes to understand a large codebase makes it a bit difficult for new devs to find their way around; but on the other hand the "culture" among the devs is very welcoming and supporting to people who are serious to contribute.

Reviewer: See comment about the entire section, but this could be a good place where to link to the resources that we actually have. The dev docs here for example.

Author: I added a link to the *Developer Guides* like so:

We have [developer documentation](https://www.getmonero.org/resources/developer-guides/), but unfortunately there is nothing that documents ...

> You find the project [here on GitHub](https://github.com/monero-project). The most important repository is the one with the core software, e.g. the Monero *daemon* that spans up the network and manages the blockchain, see [here](https://github.com/monero-project/monero). The "place to be" for Monero devs and dev-related discussion is the Libera.chat IRC channel *#monero-dev*.

Reviewer: ```suggestion
You find the project [here on GitHub](https://github.com/monero-project). The most important repository is the one with the core software, e.g. the Monero *daemon* that spans up the network and manages the blockchain, see [the repository](https://github.com/monero-project/monero). The "place to be" for Monero devs and dev-related discussion is the Libera.chat IRC channel *#monero-dev*.
```

I would also replace the last sentence with a link to the hangouts page, where people can find links to the room (and not only irc, but matrix as well)

Author: Thanks, didn't know about that nice *Hangouts* page yet. For the last sentence I arrived at:

You find the "places to be" like the dev-oriented Libera.chat IRC channel *#monero-dev* on the ["Hangouts" page](https://www.getmonero.org/community/hangouts/).

### Seraphis and Jamtis

> A group of Monero devs started to work on two brand-new and sizeable new technologies called *Seraphis* and *Jamtis* which will improve the privacy of Monero transactions further and will solve many problems regarding current functionality. Those devs get support from the inventor of Seraphis, UkoeHB, and the inventor of Jamtis, Tevador. The resulting cryptocurrency will be quite different from today's Monero, although existing users will be offered a seamless upgrade path, and funds from all the way back to 2014 remain fully spendable indefinitely.

Reviewer 2: ```suggestion
A group of Monero devs started to work on two brand-new and sizeable new technologies called *Seraphis* and *Jamtis* which will improve the privacy of Monero transactions further and will solve many problems regarding current functionality. Those devs get support from the inventor of Seraphis, [UkoeHB](https://github.com/UkoeHB), and the inventor of Jamtis, [Tevador](https://github.com/tevador). The resulting cryptocurrency will be quite different from today's Monero, although existing users will be offered a seamless upgrade path, and funds from all the way back to 2014 remain fully spendable indefinitely.
```

Author: Good idea. Added the links to UkoeHB's and Tevador's GitHubs and to the project wiki.

> Over the course of about one year UkoeHB built a library that implements all the cryptography behind Seraphis and e.g. contains code to create transactions. You find it [here](https://github.com/UkoeHB/monero/tree/seraphis_lib/src/seraphis).

Reviewer: ```suggestion
Over the course of about one year UkoeHB built a library that implements all the cryptography behind Seraphis and contains code to create transactions. You find it [on UkoeHB's repository](https://github.com/UkoeHB/monero/tree/seraphis_lib/src/seraphis).
```

Author: Changed as proposed.

Work has now started on the so-called *Seraphis wallet* that will make heavy use of the Seraphis library. *wallet* in this context does not mean an app that end users run to manage their XMR, but that component of the Monero core code that such apps will build on. The main functionality of that component: Reading and writing wallet files that contain keys and all owned outputs, scanning the blockchain for new incoming transactions, building new transactions and submitting them to the network.

The current plan is to fully replace the current component with this role called *wallet2*. This in turn will make many adjustments in the rest of the Monero core codebase necessary, and of course reworking all wallet apps to use the new component.

> If everything works out Seraphis and Jamtis may go into service with a hardfork that some still very rough estimates see happening in about 2 years.

Reviewer: This could be removed as it is a repetion of https://github.com/monero-project/monero-site/pull/2119/files#diff-935e078dfa02cd893c4720a65dfa9998d664cf2518bbd175370fda911c89b8beR13

Author: I removed the "2 years" sentence. The technically quite important detail about wanting to replace *wallet2* is not yet mentioned in the overview and might be interesting for devs.

> ### Financial Matters

Reviewer: Personally i think this section could be much shorter, with only a mention to the funding possibility and then let the website of each funding method provide the details.

Author: The same idea to directly state my info without forcing people to follow links, "advert style", would apply here as well.

> There is no company behind Monero, it has no employees. Devs seeking more or less conventional employment best look elsewhere. No venture capital companies engage in funding any work either. Monero runs on contributions from people that basically work "for free", and on donations from the broader community to finance larger tasks.

Reviewer: "There is no company behind Monero" -> "There is no company behind the Monero Project"
"No venture capital companies engage in funding any work either" <- We cannot be sure about this,probably better remove it.

Author: Added "Project". In the "advert style" I tried to find a formulation that still mentions what I see as a positive point, but takes into account the uncertainty that you rightfully point out:

 It does not actively seek venture capital either.

> There is the *CCS*, the *Community Crowdfunding System*. Its home is [here](https://ccs.getmonero.org/) on the Monero website.

Reviewer: ```suggestion
One of these funding methods is the *CCS*, the *Community Crowdfunding System*. You can find more details [in the website](https://ccs.getmonero.org/).
```

Author: Copied your proposed sentence.

Basically anybody can write a proposal for it, and if it gets positive feedback it moves to the funding stage where of course success depends on whether donators will find the proposed work and goal worthwhile.

Realistically however it's pretty hard to get a proposal accepted if you are all new to Monero, nobody in the community knows you yet and you have never contributed anything so far. Much is about trust and demonstrated abilities, as it typically is the case in open source projects.

> The CCS has a long history of successfully funded and executed tasks.

Reviewer: I wouldn't go too much into details about the functioning of the ccs in this post.

> A second option for funding Monero related work is the [MAGIC Monero Fund](https://magicgrants.org/funds/monero). Its nature is quite different from the CCS as it's a USA-registered 501(c)(3) charity. This enables tax-deductible (for citizens of the USA) donations in a variety of methods, and the option of USD-denominated payouts. There is also an additional requirement that fund recipients must verify their identity. The CCS allows its fund recipients to remain anonymous, yet only handles XMR.

Reviewer: ```suggestion
Another option for funding Monero related work is the [MAGIC Monero Fund](https://magicgrants.org/funds/monero). Its nature is quite different from the CCS as it's a USA-registered 501(c)(3) charity. This enables tax-deductible (for citizens of the USA) donations in a variety of methods, and the option of USD-denominated payouts. There is also an additional requirement that fund recipients must verify their identity.
```

Author: Adopted the "Another" improvement.

Regarding dropping the last sentence I hesitate, because I got the whole paragraph from somebody directly involved with MAGIC ... and the explicit comparison does not look too bad to me either.

### Summary

> There will be a lot of interesting work to do for devs that want to join the Monero project, for a long time, even after the hardfork to Seraphis and Jamtis in a couple of years, basically starting anytime. It may all look a bit unusual as a "place" to work at, and "get rich quick" isn't, but on the other hand think about it: Where else can you make a difference for a multi-billion-dollar venture that may become important in the future and make a difference, whoever you are, from wherever you are, regardless of education, only based on your knowledge and your willingness to make useful contributions?

Reviewer: This seems to repeat the points said elsewhere, but there is already a summary-tldr with similar content. Maybe better remove or change title and scope?

Author: I changed the title to *Conclusion*, and took out the repeated mention of Seraphis and Jamtis to shorten a bit.
